### PR TITLE
Add image logging for semantic segmentation training

### PR DESCRIPTION
**Pull request by mrpositron**
## What has changed and why?

Adds example image logging for the semantic segmentation task, mirroring the existing image logging behavior in other tasks.

`training_step` now logs a label grid for the first 3 steps (rank 0 only), showing each image with its ground-truth semantic mask overlaid on the input image.

`validation_step` now logs two grids for the first 3 steps: a ground-truth label grid and a prediction grid, showing predicted semantic masks overlaid on the input image.

Supporting infrastructure added and refined:
- `_visualize/semantic_segmentation.py` - semantic segmentation label and prediction plotting now use the model's `included_classes` mapping together with `image_normalize`, so rendered example images use the correct class names and denormalized image inputs.
- `_task_models/dinov2_eomt_semantic_segmentation/task_model.py` and `_task_models/dinov3_eomt_semantic_segmentation/task_model.py` - added `included_classes` mappings for internal class IDs, including support for the ignore class so legends can label ignored pixels correctly.
- `_task_models/dinov2_eomt_semantic_segmentation/train_model.py` and `_task_models/dinov3_eomt_semantic_segmentation/train_model.py` - training and validation steps now return label/prediction example images for early steps.
- `_visualize/utils.py` - shared helpers were added for building mask overlays, drawing mask contours, and generating legend entries for the classes present in a mask.
- `_visualize/image_classification.py` - updated to call the shared legend helper explicitly in text-only mode.

Example of a groud truth label grid:

image

## How has it been tested?

Focused visualization coverage was added in:
- `tests/_visualize/test_semantic_segmentation.py`
- `tests/_visualize/test_utils.py`

These tests cover:
- Grid output is capped at `max_images`
- Label and prediction rendering works with and without `image_normalize`
- Denormalization uses the configured mean/std values correctly
- `alpha=1.0` fully applies mask colors to the rendered image
- Class boundaries are rendered as black contours
- Legends are only shown for classes present in `included_classes`
- Mask overlay generation returns correctly sized RGB images
- Legend rendering supports colored entries, empty labels, and validation of mismatched label/color lengths
- Legend entries are generated in class-id order and skip unknown classes

I also updated semantic-segmentation train-task tests in `tests/_commands/test_train_task.py` to assert that example images are written to `image_examples/` for training and validation runs.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py [hunk 1]
```diff
@@ -142,6 +147,11 @@ def __init__(
             loss_names=["loss"],
         )
 
+        # TODO(Nauryz, 04/2026): These visualization thresholds are currently
+        # hardcoded, but we may want to make them configurable in the future
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py:151]: Can you create an issue for that if not existant yet.

**Comment by gabrielfruet** [comment]
GitHub is not working properly so I'm going to comment my review here.

`(...)/_visualize/utils.py:_render_grid` -> Doesn't account for the case where the list of images is empty. Maybe just adding in the docstring that the list shall not empty would be a good idea.

**Comment by gabrielfruet** [comment]
src/lightly_train/_task_models/*/train_model.py
 
Why the logging is gated by `step < 3`  ?

**Comment by gabrielfruet** [comment]
It's looking good! I will just ask Lionel if we are over-testing in `test_utils.py`, I don't know the guidelines

File: tests/_visualize/test_utils.py [hunk 2]
```diff
@@ -11,179 +11,409 @@
...
-
-    def test__get_class_color_large_class_id(self) -> None:
-        r, g, b = utils._get_class_color(10_000)
+def _non_white_bbox(image: PILImage) -> tuple[int, int, int, int] | None:
+    """Bounding box of pixels that differ from pure white, or None if all white."""
+    white = Image.new(image.mode, image.size, _WHITE_PIXEL)
+    return ImageChops.difference(image, white).getbbox()
+
+
+def test__cxcywh_to_xyxy__center_box() -> None:
```
Comment (yutong-xiang-97) [inline tests/_visualize/test_utils.py]: The following 4 tests can be parameterized with `@pytest.mark.parameterize`

File: tests/_commands/test_train_task.py [hunk 3]
```diff
@@ -638,6 +645,13 @@ def test_train_semantic_segmentation__export(
     for key in ckpt_model_state_dict.keys():
         assert torch.equal(ckpt_model_state_dict[key], exported_model_state_dict[key])
 
+    # Check that example images were logged for training and validation.
+    image_examples_dir = out / "image_examples"
+    assert image_examples_dir.exists()
```
Comment (yutong-xiang-97) [inline tests/_commands/test_train_task.py]: This only needs to be validated once in `test_train_semantic_segmentation`.

File: tests/_visualize/test_semantic_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+def test_plot_semantic_segmentation_predictions__no_image_normalize_skips_denormalization() -> (
+    None
+):
+    # Without image_normalize the image tensor passes through unchanged.
+    # alpha=0 disables the overlay so 0.4 -> 102 is preserved.
+    batch = _make_batch(
+        image=torch.full((1, 3, 32, 32), 0.4),
+        mask=torch.zeros(1, 32, 32, dtype=torch.long),
+    )
+    logits = [torch.zeros(2, 32, 32)]
```
Comment (yutong-xiang-97) [inline tests/_visualize/test_semantic_segmentation.py]: From Claude Code:

The argmax-over-logits behavior (the only thing that distinguishes predictions from labels) is never actually verified. Add a test with non-uniform logits where the argmax class differs per region, and assert the rendered overlay color matches that class.

File: tests/_visualize/test_semantic_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        included_classes={1: "dog"},
+        max_images=1,
+        image_normalize=None,
+        alpha=0.0,
+    )
+    assert _has_legend(with_known_class)
+    assert not _has_legend(with_unknown_class)
+
+
+def test_plot_semantic_segmentation_predictions__grid_caps_at_max_images() -> None:
```
Comment (yutong-xiang-97) [inline tests/_visualize/test_semantic_segmentation.py]: From Claude Code:

- `__grid_caps_at_max_images`
- `__no_image_normalize_skips_denormalization`
- `__image_normalize_denormalizes`

Each one is an exact structural mirror of the corresponding `labels` test above it. They build the same batch, pass the same `image_normalize` arg, and assert the same pixel value at the same coordinate. The only difference is the function under test.

That looks like coverage but it's actually testing the **shared dependencies**, not the function itself. Both `plot_semantic_segmentation_labels` and `plot_semantic_segmentation_predictions` call the same `_denormalize_image` and `_render_grid` helpers — and those helpers already have direct unit tests in test_utils.py:172-213 (denormalize) and test_utils.py:65-107 (grid). So the predictions-side denormalize tests prove "predictions also calls `_denormalize_image`" — a wiring check, not a behavior check.

Two reasonable ways to cut this:

**Option A — keep one wiring test per entry point.** Drop `__no_image_normalize_skips_denormalization` from the predictions side and keep only `__image_normalize_denormalizes`. The single test confirms predictions wires denormalize correctly; the unit tests in `test_utils.py` cover the actual math. Same for labels. You go from 4 tests → 2.

**Option B — drop all three predictions-side mirror tests.** Rely on the labels-side tests + the helper unit tests to cover denormalize/grid behavior. Replace them with **one** test that's actually predictions-specific: feed non-uniform logits where `argmax` picks class 1 in some region and class 0 in another, and assert the overlay colors match. Net is 3 tests removed, 1 added — and the new one covers the only behavior unique to `predictions` that currently isn't tested anywhere.

I'd lean toward **B**. The denormalize/grid mirror tests will pass even if argmax is broken or the wrong logit index is used, so they don't earn their slot. The argmax test would catch a real bug class.

**Review by yutong-xiang-97 — COMMENTED** [review]
Add some review to the tests. I'd leave the final approval to @gabrielfruet

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py [hunk 6]
```diff
@@ -245,6 +251,12 @@ def __init__(
             self, hooks.criterion_empty_weight_reinit_hook
         )
 
+        # TODO(Nauryz, 04/2026): These visualization thresholds are currently
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py:254]: Make sure to add an issue for this.

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py [hunk 7]
```diff
@@ -314,11 +326,20 @@ def training_step(
                 current_iter=step,
                 final_iter=no_auto(self.model_args.attn_mask_annealing_steps_end)[i],
             )
-
+        label_image: PILImage | None = None
+        if step < 3 and fabric.global_rank == 0:
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py:330]: I guess the `3` should equally be configurable with the `viz_max_images` and `viz_alpha`.

File: tests/_visualize/test_utils.py (hunk 2 above)
Comment (gabrielfruet) [reply tests/_visualize/test_utils.py]: @mrpositron You removed these tests? or you collapsed then to `test__cxcywh_to_xyxy` ?

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py [hunk 8]
```diff
@@ -164,7 +175,21 @@ def training_step(
         if self.metric_args.train:
             self.train_metrics.update_with_predictions(logits.argmax(dim=1), masks)
 
-        return TaskStepResult(loss=loss, log_dict={}, metrics=self.train_metrics)
+        label_image: PILImage | None = None
+        if step < 3 and fabric.global_rank == 0:
```
Comment (gabrielfruet) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py:179]: Add a TODO here to remember to add the step also to logger_args

Comment (gabrielfruet) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py:179]: Also in other related places that share this same logic.

**Review by gabrielfruet — CHANGES_REQUESTED** [review]

**Comment by gabrielfruet** [comment]
Just minor changes related to TODOs. Looking good!

**Review by liopeer — COMMENTED** [review]
LGTM from my side. Leave final approval to @yutong-xiang-97 !

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py (hunk 1 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py:151]: This would be in the cleanup issue.

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py (hunk 6 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py:254]: Same as above.

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py (hunk 7 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/train_model.py:330]: Ultralytics only logs the first 3 examples but we can ofc do it if it is easy.

File: tests/_visualize/test_utils.py (hunk 2 above)
Comment (mrpositron) [reply tests/_visualize/test_utils.py]: Yes, I have removed these tests

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py (hunk 8 above)
Comment (mrpositron) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/train_model.py:179]: There will be a cleanup PR later; I will update them there.

**Review by gabrielfruet — APPROVED** [review]
